src/CoreFoundation/CFProxySupport.cs

```diff
@@ -579,62 +581,241 @@ public static CFProxySettings GetSystemProxySettings ()
+
+		static CFProxy[] ParseProxies (IntPtr proxyList) {
+			CFProxy[] proxies = null;
+			if (proxyList != IntPtr.Zero) {
+				using (var array = new CFArray (proxyList, false)){
+					proxies = new CFProxy [array.Count];
+					for (int i = 0; i < proxies.Length; i++) {
+						var dict = new NSDictionary (array.GetValue (i));
+						proxies[i] = new CFProxy (dict);
+					}
+				}
+			}
+			return proxies;
+		}
+
+		static async Task<(CFProxy[] proxies, NSError error)> ExecutePacCFRunLoopSourceAsync (CreatePACCFRunLoopSource factory, CancellationToken cancellationToken)
+		{
+			CFProxy[] proxies = null;
+			NSError outError = null;
+			await Task.Run (() => {
+				// we need the runloop of THIS thread, so it is important to get it in the correct context
+				var runLoop = CFRunLoop.Current;
+				CFProxyAutoConfigurationResultCallbackInternal cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+					proxies = ParseProxies (proxyList);
+					if (error != IntPtr.Zero)
+						outError = new NSError (error);
+					runLoop.Stop ();
+				};
+				var clientContext = new CFStreamClientContext ();
+				using (var loopSource = new CFRunLoopSource (factory (cb, ref clientContext)))
+				using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+					runLoop.AddSource (loopSource, mode);
+					// blocks until stop is called, will be done in the cb set previously
+					runLoop.RunInMode (mode, double.MaxValue, false);
+					runLoop.RemoveSource (loopSource, mode);
+				}
+			}, cancellationToken).ConfigureAwait (false);
+
+			return (proxies: proxies, error: outError);
+		}
+
+		static CFProxy[] ExecutePacCFRunLoopSourceBlocking (CreatePACCFRunLoopSource factory, out NSError outError)
+		{
+			outError = null;
+			IntPtr cbError = IntPtr.Zero;
+			CFProxy[] proxies = null;
+
+			var runLoop = CFRunLoop.Current;
+			CFProxyAutoConfigurationResultCallbackInternal cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+				proxies = ParseProxies (proxyList);
+				cbError = error;
+				runLoop.Stop ();
+			};
+
+			var clientContext = new CFStreamClientContext ();
+			using (var loopSource = new CFRunLoopSource (factory (cb, ref clientContext)))
+			using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+				runLoop.AddSource (loopSource, mode);
+				runLoop.RunInMode (mode, double.MaxValue, false);
+				runLoop.RemoveSource (loopSource, mode);
+				if (cbError != IntPtr.Zero)
+					outError = new NSError (cbError);
+				return proxies;
+			}
+		}
+
+		static CFRunLoopSource ExecutePacCFRunLoopSourceBlocking (CreatePACCFRunLoopSource factory, CFProxyAutoConfigurationResultCallback clientCb, CFStreamClientContext clientContext)
+		{
+			CFProxyAutoConfigurationResultCallbackInternal cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+				var proxies = ParseProxies (proxyList);
+				clientCb ((client != IntPtr.Zero)? null : new NSObject (client), proxies, (error == IntPtr.Zero) ? null : new NSError (error));
```

Comment:
`Runtime.GetNSObject` is likely better suited (for both `client` and `error`) since it won't create a 2nd managed instance over the same native one (if that's possible, not clear to me here)